File: 0067-add-binary/0067-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        
        i = len(a) - 1
        j = len(b) - 1
        carry = 0
        resStr = ""
        
        while i >= 0 or j >= 0 or carry:
            #Convert int
            #2. Bounds check?

            if i < 0:
                bitA = 0
            else:
                bitA = int(a[i])

            if j < 0:
                bitB = 0
            else:
                bitB = int(b[j])
            
            #0, 1, 2, 3
            total = bitA + bitB + carry

            if total == 0:
                resStr += str(total)
                carry = 0
            elif total == 1:
                resStr += str(total)
                carry = 0
            elif total == 2:
                resStr += str(0)
                carry = 1
            elif total == 3:
                resStr += str(1)
                carry = 1

            i -= 1
            j -= 1

        return resStr[::-1]
        

File: 0067-add-binary/0067-add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) 
    {
        int i = a.size() - 1;
        int j = b.size() - 1;
        int carry = 0;
        string res = "";

        while (i >= 0 || j >= 0 || carry)
        {
            int bitA = (i >= 0) ? (a[i] - '0') : 0;
            int bitB = (j >= 0) ? (b[j] - '0') : 0;
            int alphaBit = 0;

            if (bitA == 1 && bitB == 1 && carry == 0)
            {
                alphaBit = 0;    
                carry = 1;
            }

            else if (bitA == 1 && bitB == 1 && carry == 1)
            {
                alphaBit = 1;
                carry = 1;
            }

            else if (bitA == 0 && bitB == 1 && carry == 1)
            {
                alphaBit = 0;
                carry = 1;
            }

            else if (bitA == 0 && bitB == 1 && carry == 0)
            {
                alphaBit = 1;
                carry = 0;
            }

            else if (bitA == 1 && bitB == 0 && carry == 1)
            {
                alphaBit = 0;
                carry = 1;
            }

            else if (bitA == 1 && bitB == 0 && carry == 0)
            {
                alphaBit = 1;
                carry = 0;
            }

            else if (bitA == 0 && bitB == 0 && carry == 0)
            {
                alphaBit = 0;
                carry = 0;
            }

            else //bitA == 0 && bitB == 0 && carry == 1
            {
                alphaBit = 1;
                carry = 0;
            }

            res.push_back(alphaBit + '0');

            i -= 1;
            j -= 1;

        }

        //Reverse
        reverse(res.begin(), res.end());

        return res;
    }

private:

};